### PR TITLE
launcher: Use actual executable path instead of `argv[0]`

### LINKED ISSUE #641: `jazzer` fails to find `jazzer_standalone.jar` when `jazzer` is not run from the directory containing it
I'm running `jazzer` like this:
`fuzz.sh`:
```bash
#!/bin/bash

export PATH=$PATH:/home/sg/dev/jazzer-new/
whereis jazzer
./gradlew jar && jazzer --cp=lib/build/libs/lib.jar --target_class=javac.fuzzing.invalid.classfile.FuzzingTest
```
and then in a terminal: `./fuzz.sh`.
Note that `fuzz.sh` is located in a different directory from the `jazzer` binary.

**Expected Result**: Fuzzing :tada: 
**Actual Result**:
> ERROR: Could not find jazzer_standalone.jar. Please provide the pathname via the --agent_path flag.

If I understand [this comment](https://github.com/CodeIntelligenceTesting/jazzer/blob/dcd70dd5977854b51e46b3d8fffc815fb31e2451/launcher/jvm_tooling.cpp#L85) correctly, this _should_ work:
>  // If the agent is not in the bazel path we look next to the jazzer binary.
However, looking at the `strace` output, it appears that it merely looks at the CWD which might be different from the location of the `jazzer` binary.

`strace` output:
<details>

```shell
execve("/home/sg/dev/jazzer-new/jazzer", ["jazzer", "--cp=lib/build/libs/lib.jar", "--target_class=javac.fuzzing.inv"...], 0x7ffce1afb9a0 /* 52 vars */) = 0
brk(NULL)                               = 0xd87000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=215988, ...}) = 0
mmap(NULL, 215988, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6559202000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libm.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\322\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=1321344, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6559200000
mmap(NULL, 1323280, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f65590bc000
mmap(0x7f65590c9000, 630784, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xd000) = 0x7f65590c9000
mmap(0x7f6559163000, 634880, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xa7000) = 0x7f6559163000
mmap(0x7f65591fe000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x141000) = 0x7f65591fe000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0000\21\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=18688, ...}) = 0
mmap(NULL, 20752, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f65590b6000
mmap(0x7f65590b7000, 8192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1000) = 0x7f65590b7000
mmap(0x7f65590b9000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f65590b9000
mmap(0x7f65590ba000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f65590ba000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0 l\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=149520, ...}) = 0
mmap(NULL, 136304, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6559094000
mmap(0x7f655909a000, 65536, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6000) = 0x7f655909a000
mmap(0x7f65590aa000, 24576, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x16000) = 0x7f65590aa000
mmap(0x7f65590b0000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1b000) = 0x7f65590b0000
mmap(0x7f65590b2000, 13424, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f65590b2000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@>\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1905632, ...}) = 0
mmap(NULL, 1918592, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6558ebf000
mmap(0x7f6558ee1000, 1417216, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7f6558ee1000
mmap(0x7f655903b000, 323584, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17c000) = 0x7f655903b000
mmap(0x7f655908a000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1ca000) = 0x7f655908a000
mmap(0x7f6559090000, 13952, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f6559090000
close(3)                                = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6558ebd000
arch_prctl(ARCH_SET_FS, 0x7f6558ebdb80) = 0
mprotect(0x7f655908a000, 16384, PROT_READ) = 0
mprotect(0x7f65590b0000, 4096, PROT_READ) = 0
mprotect(0x7f65590ba000, 4096, PROT_READ) = 0
mprotect(0x7f65591fe000, 4096, PROT_READ) = 0
mprotect(0x270000, 24576, PROT_READ)    = 0
mprotect(0x7f6559261000, 4096, PROT_READ) = 0
munmap(0x7f6559202000, 215988)          = 0
set_tid_address(0x7f6558ebde50)         = 478185
set_robust_list(0x7f6558ebde60, 24)     = 0
rt_sigaction(SIGRTMIN, {sa_handler=0x7f655909a690, sa_mask=[], sa_flags=SA_RESTORER|SA_SIGINFO, sa_restorer=0x7f65590a7140}, NULL, 8) = 0
rt_sigaction(SIGRT_1, {sa_handler=0x7f655909a730, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_SIGINFO, sa_restorer=0x7f65590a7140}, NULL, 8) = 0
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
gettid()                                = 478185
gettid()                                = 478185
gettid()                                = 478185
gettid()                                = 478185
brk(NULL)                               = 0xd87000
brk(0xda8000)                           = 0xda8000
openat(AT_FDCWD, "", O_RDONLY)          = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
stat("", 0x7ffe6fb4ef50)                = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "", O_RDONLY)          = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
stat("", 0x7ffe6fb4ef60)                = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "jazzer.runfiles/MANIFEST", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
stat("jazzer.runfiles", 0x7ffe6fb4ef60) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "jazzer.runfiles_manifest", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "", O_RDONLY)          = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
stat("", 0x7ffe6fb4ed90)                = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "jazzer.runfiles/MANIFEST", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
stat("jazzer.runfiles", 0x7ffe6fb4ed90) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "jazzer.runfiles_manifest", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/jazzer_standalone.jar", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
write(2, "ERROR: Could not find ", 22ERROR: Could not find )  = 22
futex(0x276174, FUTEX_WAKE_PRIVATE, 2147483647) = 0
write(2, "jazzer_standalone.jar", 21jazzer_standalone.jar)   = 21
write(2, ". Please provide the pathname vi"..., 56. Please provide the pathname via the --agent_path flag.) = 56
write(2, "\n", 1
)                       = 1
exit_group(1)                           = ?
+++ exited with 1 +++

```
</details>

### PR DESCRIPTION
**Pull request by fmeum**
When Jazzer is executed from `PATH`, `argv[0]` will just be `jazzer`, which we can't find `jazzer_standalone.jar` relative to.

Also fix `argv[0]` in the case where no shell launcher is generated.

Fixes #641

### CODE REVIEW

File: launcher/jvm_tooling.cpp [hunk 1]
```diff
@@ -50,6 +57,32 @@ constexpr auto kJazzerBazelRunfilesPath =
...
+// this path can always be used to locate the Jazzer JAR next to it, even when
+// Jazzer is executed from PATH.
+std::string getExecutablePath() {
+#if defined(__APPLE__)
+  char buf[655536];
+  uint32_t buf_size = sizeof(buf);
+  if (!_NSGetExecutablePath(buf, &buf_size)) {
+    return "";
+  }
+  return {buf};
```
Comment (intrigus-lgtm) [inline launcher/jvm_tooling.cpp:70]: I'm not a C++ expert, but isn't this returning stack memory to the caller?
I.e. the caller will be using memory after it has been "freed"?

Comment (fmeum) [reply launcher/jvm_tooling.cpp:70]: Thanks for taking a look. Note the curly braces, which end up passing `buf` to constructor (5) at https://en.cppreference.com/w/cpp/string/basic_string/basic_string - that constructor copies the null-terminated string.

**Review by bertschneider — APPROVED** [review]
Looks good. I wonder if we could add a test for this.

**Comment by fmeum** [comment]
The code path is also used in all Bazel tests now as we pass the result of the new function to the C++ runfiles library. But I will try to add coverage for `PATH` invocation specifically.

**Pull request merged**